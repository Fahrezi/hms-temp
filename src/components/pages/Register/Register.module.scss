.register {
  & {
    min-height: 100vh;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: #f3f4f6;
  }

  &__card {
    background: white;
    padding: 2rem;
    border-radius: 0.5rem;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.1);
    width: 100%;
    max-width: 400px;

    &--title {
      font-size: 1.5rem;
      font-weight: 600;
      margin-bottom: 1.5rem;
      text-align: center;
    }
  }

  &__field {
    margin-bottom: 1rem;

    &--label {
      display: block;
      margin-bottom: 0.5rem;
      font-weight: 500;
      color: #666;
    }

    &--input {
      width: 100%;
      padding: 0.5rem;
      border: 1px solid #ddd;
      border-radius: 0.25rem;

      &:focus {
        outline: none;
        border-color: #2563eb;
      }
    }
  }

  &__button {
    width: 100%;
    padding: 0.75rem;
    background-color: #2563eb;
    color: white;
    border: none;
    border-radius: 0.25rem;
    font-weight: 500;
    cursor: pointer;
    transition: background-color 0.2s;

    &:hover {
      background-color: #1d4ed8;
    }

    &:focus {
      outline: none;
      box-shadow: 0 0 0 2px rgba(37, 99, 235, 0.5);
    }
  }
}
