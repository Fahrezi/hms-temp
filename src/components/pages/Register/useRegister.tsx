import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { z } from 'zod';

import { registerRequest } from '@/services/auth.service';

const registerSchema = z.object({
  email: z.string().email('Invalid email address'),
  password: z.string().min(6, 'Password must be at least 6 characters'),
});

type RegisterFormInputs = z.infer<typeof registerSchema>;

export const useRegister = () => {
  const navigate = useNavigate();
  const registerMutation = registerRequest();

  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
  } = useForm<RegisterFormInputs>({
    resolver: zodResolver(registerSchema),
    mode: 'onBlur',
  });

  const onSubmit = async (data: RegisterFormInputs) => {
    try {
      await registerMutation.mutateAsync(data);
      alert('Registration successful!');
      void navigate('/login');
    } catch (error) {
      console.error('Registration error:', error);
    }
  };

  return {
    register,
    handleSubmit,
    onSubmit,
    errors,
    isSubmitting: isSubmitting || registerMutation.isPending,
  };
};
