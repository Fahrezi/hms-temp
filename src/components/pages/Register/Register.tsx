import { useRegister } from './useRegister';

import styles from './Register.module.scss';

const Register = () => {
  const { register, handleSubmit, onSubmit, errors, isSubmitting } = useRegister();

  return (
    <div className={styles.register}>
      <div className={styles.register__card}>
        <header>
          <h1 className={styles['register__card--title']}>Register</h1>
        </header>
        <form onSubmit={(e) => void handleSubmit(onSubmit)(e)} role='form'>
          <div className={styles.register__field}>
            <label htmlFor='email' className={styles['register__field--label']}>
              Email
            </label>
            <input
              id='email'
              type='email'
              {...register('email')}
              className={styles['register__field--input']}
              disabled={isSubmitting}
            />
            {errors.email && <p className={styles['register__error']}>{errors.email.message}</p>}
          </div>

          <div className={styles.register__field}>
            <label htmlFor='password' className={styles['register__field--label']}>
              Password
            </label>
            <input
              id='password'
              type='password'
              {...register('password')}
              className={styles['register__field--input']}
              disabled={isSubmitting}
            />
            {errors.password && <p className={styles['register__error']}>{errors.password.message}</p>}
          </div>

          <button
            type='submit'
            className={styles.register__button}
            disabled={isSubmitting}
            aria-label={isSubmitting ? 'Registering...' : 'Register'}
          >
            {isSubmitting ? 'Registering...' : 'Register'}
          </button>
        </form>
        <p className={styles.register__login}>
          Already have an account? <a href='/login'>Login here</a>
        </p>
      </div>
    </div>
  );
};

export default Register;
