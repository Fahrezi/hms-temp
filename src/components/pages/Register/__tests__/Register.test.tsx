import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import Register from '@/components/pages/Register';

const queryClient = new QueryClient();

describe('Register page', () => {
  it('renders the Register component correctly', () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <Register />
        </QueryClientProvider>
      </MemoryRouter>,
    );

    expect(screen.getByRole('heading', { name: 'Register' })).toBeInTheDocument();
    expect(screen.getByRole('form')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Register' })).toBeInTheDocument();
  });

  it('disables the button when submitting', () => {
    render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <Register />
        </QueryClientProvider>
      </MemoryRouter>,
    );

    const button = screen.getByRole('button', { name: 'Register' });
    fireEvent.click(button);

    expect(button).toBeDisabled();
    expect(screen.getByRole('button', { name: 'Registering...' })).toBeInTheDocument();
  });

  it('matches the Register component snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <QueryClientProvider client={queryClient}>
          <Register />
        </QueryClientProvider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });
});
