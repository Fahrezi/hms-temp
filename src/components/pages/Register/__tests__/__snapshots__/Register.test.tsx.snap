// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Register page > matches the Register component snapshot 1`] = `
<div>
  <div
    class="_register_5257f9"
  >
    <div
      class="_register__card_5257f9"
    >
      <header>
        <h1
          class="_register__card--title_5257f9"
        >
          Register
        </h1>
      </header>
      <form
        role="form"
      >
        <div
          class="_register__field_5257f9"
        >
          <label
            class="_register__field--label_5257f9"
            for="email"
          >
            Email
          </label>
          <input
            class="_register__field--input_5257f9"
            id="email"
            name="email"
            type="email"
          />
        </div>
        <div
          class="_register__field_5257f9"
        >
          <label
            class="_register__field--label_5257f9"
            for="password"
          >
            Password
          </label>
          <input
            class="_register__field--input_5257f9"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <button
          aria-label="Register"
          class="_register__button_5257f9"
          type="submit"
        >
          Register
        </button>
      </form>
      <p
        class="_register__login_5257f9"
      >
        Already have an account? 
        <a
          href="/login"
        >
          Login here
        </a>
      </p>
    </div>
  </div>
</div>
`;
